### PR TITLE
RND Rework [Death to Techweb]

### PR DESCRIPTION
**Pull request by EmoGarbage404**
## About the PR

Based off of [this design doc.](https://hackmd.io/@ss14-design/Sk1VIWjni)

### The TL;DR
All technologies are now sorted into (as of right now) 4 categories called "Disciplines." Each discipline roughly corresponds to a gameplay loop (engie/salvage, medical, science, service) and contains a variety of technology that unlocks lathe recipes.

Each tech, like before, has a cost that must be paid to unlock it. However, technologies aren't unlocked linearly. Each tech is assigned a "tier," and researching a significant portion (75% as of right now) of a discipline's techs of a tier will unlock the next one.

Furthermore, tier 3 is a "lockout tech." Unlocking a tier 3 tech will set it as your main discipline, preventing you from researching other tier 3 techs of other disciplines.

The last point of note is the card system. Instead of having all available techs be researchable, you have to pick from 4, 1 from each discipline, to unlock. After unlocking one, 4 new cards will be drawn that you have to choose from.

**Media**

image
image
image
image

- [x] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

**Changelog**

:cl:
- add: The autolathe catalog has been expanded.
- add: You can now see what recipes a technology unlocks.
- tweak: The research console can now only select technologies from the specific options given to the player.
- tweak: Technologies are now sorted into disciplines. Unlock all the techs in a discipline to unlock higher tiered tech.

### CODE REVIEW

**Comment by mirrorcult** [comment]
for the record this design approved by me at least

**Comment by forgotmyotheraccount** [comment]
My absolute favourite part about this rework is the fact that it actually tells you which items each technology unlocks, as someone who's new to the Science tech tree it's been infuriating trying to find out what I need to research.

Admittedly it's also been difficult to figure out what I've already unlocked. Is there anyway to view the items of previously unlocked technologies, e.g. mousing over them in the tab?

**Comment by EmoGarbage404** [comment]
> Admittedly it's also been difficult to figure out what I've already unlocked. Is there anyway to view the items of previously unlocked technologies, e.g. mousing over them in the tab?

that functionality is in the ui. Mousing over any of the available/already unlocked researches will show a tooltip with all of the recipes they unlock.

**Comment by EmoGarbage404** [comment]
Ready for review
econonygood

**Comment by Anzarot121** [comment]
I have a question. Is specialization tied to a game round or to RnD server?

**Comment by EmoGarbage404** [comment]
> I have a question. Is specialization tied to a game round or to RnD server?

It's per-server.

File: Resources/Prototypes/Research/technology.yml [hunk 1]
```diff
@@ -0,0 +1,621 @@
...
+  discipline: Biochemical
+  tier: 2
+  cost: 10000
+  recipeUnlocks:
+  - HandheldCrewMonitor
+
+# Tier 3
+
+- type: technology
+  id: Cloning
```
Comment (0x6273) [inline Resources/Prototypes/Research/technology.yml:290]: Cloning being a T3 tech seems kinda strange, since cloning is mapped in and available on roundstart. I think this will lead to a lot of early shuttle calls if cloning gets bombed and sci already unlocked a T3 from a different discipline.

Comment (deltanedas) [reply Resources/Prototypes/Research/technology.yml:290]: arguments against:
- cloning is rarely needed with good med staff since defibs cover 90% of cases
- cloning is not available roundstart on at least 1 map so yeah

**Comment by ThunderBear2006** [comment]
This looks really awesome Emo, keep up the good work!

File: Content.Client/Research/UI/TechnologyCardControl.xaml.cs [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+        RobustXamlLoader.Load(this);
+
+        var discipline = prototypeManager.Index<DisciplinePrototype>(technology.Discipline);
+        Background.ModulateSelfOverride = discipline.Color;
+
+        DisciplineTexture.Texture = spriteSys.Frame0(discipline.Icon);
+        TechnologyNameLabel.Text = Loc.GetString(technology.Name);
+        var message = new FormattedMessage();
+        message.AddMarkup($"{Loc.GetString("research-console-tier-discipline-info",
+            ("tier", technology.Tier), ("color", discipline.Color), ("discipline", Loc.GetString(discipline.Name)))}");
```
Comment (mirrorcult) [inline Content.Client/Research/UI/TechnologyCardControl.xaml.cs:27]:
```suggestion
        message.AddMarkup(Loc.GetString("research-console-tier-discipline-info",
            ("tier", technology.Tier), ("color", discipline.Color), ("discipline", Loc.GetString(discipline.Name))));
```

File: Content.Shared/Research/Systems/SharedResearchSystem.cs [hunk 3]
```diff
@@ -1,46 +1,147 @@
...
+        {
+            var filtered = availableTechnology.
+                Where(p => p.Discipline == discipline &&
+                           !component.UnlockedTechnologies.Contains(p.ID)).ToList();
+
+            if (!filtered.Any())
+                continue;
+
+            component.CurrentTechnologyCards.Add(_random.Pick(filtered).ID);
+        }
```
Comment (mirrorcult) [inline Content.Shared/Research/Systems/SharedResearchSystem.cs:44]: imo makes more sense to just shuffle `availableTechnology` then iterate over it and pick the first of each discipline thats found. `!component.UnlockedTechnologies.Contains(p.ID)` shouldnt be necessary since `IsTechnologyAvailable` already checks that

**Review by mirrorcult — CHANGES_REQUESTED** [review]

**Comment by EmoGarbage404** [comment]
review finished.

**Comment by OttoMaticode** [comment]
HUGE

great work bro

**Comment by deltanedas** [comment]
:rocket:

**Comment by DawBla** [comment]
Holy shit it's here
Good stuff

**Comment by metalgearsloth** [comment]
> Did you really have to make bluespace storages THAT cheap? Like, really?

If you want to provide useful feedback please do it without sarcasm.

**Comment by EmoGarbage404** [comment]
> Did you really have to make bluespace storages THAT cheap? Like, really? 

I dont think half a stack of plasma and steel is that cheap.

**Comment by Gerkada** [comment]
> > Did you really have to make bluespace storages THAT cheap? Like, really?
> 
> I dont think half a stack of plasma and steel is that cheap.

Considering how easy it is for cargo to farm a morbillion space credits, yes, it is cheap. I've had this convo under the Liltenhead video. The fact that it requires "a lot of research points" and "is an endgame research" are kinda bad points. Even on LRP servers with 1-2h shifts it is rly easy to speedrun the whole tech tree (artifacts have like 5 conditions to activate, and only 2 are actually hard (high pressure and radiation)). So ur average science boy will be able to get his hands on the desired tech 3 research in no time + he can get all of tier 3 research via printing disks. After that, he just goes to cargo, after they acquired more money than some countries have IRL (frezon is still easy to make), orders 10 crates of steel, plasma and plastic and mass produces OP bluespace bags. After that everyone on the station is capable of having LITERALLY EVERYTHING on them, starting from the whole freaking medbay, materials enough to build another station from scratch and ending with local armory in their pockets. And I'm not even gonna talk how OP it is for secoffs. Mfs can have 10 guns and 20 ammo boxes on them and shoot n scoop nonstop. Imagine flukies faces when they arrive at the station and every single secoff has like the whole military base in their pockets as well as 10 medkits. So I thinks it's a REALLY BAD idea to make bluespace bags so easy to make. And I'm not even going to mention MRP servers with 3h+ long shifts, lol.

**Comment by metalgearsloth** [comment]
You're just re-stating https://github.com/space-wizards/space-station-14/issues/14310

**Comment by Gerkada** [comment]
> You're just re-stating #14310

Even w/o frezon i believe it's still easy to farm credits for cargo, so my point still stands

**Comment by deltanedas** [comment]
yeah even just doing 1 (one) hazard 3-5 expedition can net you lile 10-30k maybe
convert that into mats and you get a lot of bags
this is why i think bags of holding need to be harder to make (bluespace? like on nyano (although from mining not oracle) which has it used in bags of holding)
